### PR TITLE
EVM Contract Support

### PR DESCRIPTION
**Pull request by Wizdave97**
Depends on https://github.com/polytope-labs/ismp-rs/pull/63

### CODE REVIEW

File: pallet-ismp/src/evm/abi.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            uint256 timeoutTimestamp;
+            // raw storage keys
+            bytes[] keys;
+            // height at which to read destination state machine
+            uint256 height;
+        }
+
+        struct StorageValue {
+            bytes key;
+            OptionValue value;
```
Comment (seunlanlege) [inline pallet-ismp/src/evm/abi.rs:41]:
```suggestion
            bytes value;
```

File: pallet-ismp/src/evm/handler.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        };
+        let call_data = OnAcceptCall { request: post }.encode();
+        execute_call::<T>(target_contract, call_data, gas_limit)
+    }
+
+    fn on_response(&self, response: Response) -> Result<Gas, Error> {
+        let target_contract = parse_contract_id(&response.destination_module())?;
+        let gas_limit = GasLimits::<T>::get(response.nonce())
+            .ok_or(Error::ImplementationSpecific("Gas limit not found".to_string()))?;
+        GasLimits::<T>::remove(response.nonce());
```
Comment (seunlanlege) [inline pallet-ismp/src/evm/handler.rs:62]: thought this was only for get requests?

File: pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            from: context.caller.0.to_vec(),
+            to: post_dispatch.to,
+            timeout_timestamp: u256_to_u64(post_dispatch.timeoutTimestamp)?,
+            data: post_dispatch.data,
+        };
+        handle.record_cost(cost)?;
+        match dispatcher.dispatch_request(DispatchRequest::Post(post_dispatch)) {
+            Ok(_) => {
+                let nonce = Pallet::<T>::previous_nonce();
+                GasLimits::<T>::insert(nonce, gas_limit);
```
Comment (seunlanlege) [inline pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs:65]: but this gas limit is for the destination chain

File: pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+/// Convert u256 to u64 with overflow check
+fn u256_to_u64(value: alloy_primitives::U256) -> Result<u64, PrecompileFailure> {
+    let value = U256::from_big_endian(value.to_be_bytes::<32>().as_slice());
+    for i in &value.0[1..] {
+        if *i != 0u64 {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::Other("Integer Overflow".into()),
+            })
+        }
```
Comment (seunlanlege) [inline pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs:178]: why would it overflow?

File: pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs (hunk 3 above)
Comment (Wizdave97) [reply pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs:65]: The gas limit for the destination chain is encoded with the request data as part of the `ContractData` struct.

File: pallet-ismp/src/evm/handler.rs (hunk 2 above)
Comment (seunlanlege) [reply pallet-ismp/src/evm/handler.rs:62]: hmm, let's do it this way, when dispatching requests, apps should set `gas_limit` to `max(destination_gas_limit, response_gas_limit)`

File: pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs (hunk 4 above)
Comment (Wizdave97) [reply pallet-ismp/src/evm/ismp_dispatcher_precompiles.rs:178]: it should not ideally, I'll add it as a comment instead and use the default method.

File: pallet-ismp/evm/src/handler.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            let mut total_weight_used = WeightConsumed::<T>::get();
+            let weight_limit = T::GasWeightMapping::gas_to_weight(gas_limit, true);
+            total_weight_used.weight_used = total_weight_used.weight_used + error.weight;
+            total_weight_used.weight_limit = total_weight_used.weight_limit + weight_limit;
+            WeightConsumed::<T>::put(total_weight_used);
+            // We still return ok so we can compensate for used gas only
+            Err(Error::ImplementationSpecific(
+                "Contract encountered error while executing".to_string(),
+            ))
+        }
```
Comment (seunlanlege) [inline pallet-ismp/evm/src/handler.rs:211]: prefer if we dedupe this match case.
```suggestion
let weight_used = match result {
    Ok(info) => T::GasWeightMapping::gas_to_weight(info.used_gas.standard.low_u64(), true),
    Err(error) => error.weight,
}

let mut total_weight_used = WeightConsumed::<T>::get();
let weight_limit = T::GasWeightMapping::gas_to_weight(gas_limit, true);
total_weight_used.weight_used = total_weight_used.weight_used + weight_used;
total_weight_used.weight_limit = total_weight_used.weight_limit + weight_limit;
WeightConsumed::<T>::put(total_weight_used);
```

File: pallet-ismp/src/lib.rs [hunk 6]
```diff
@@ -529,7 +548,13 @@ where
             Self::deposit_event(Event::<T>::HandlingErrors { errors })
         }
 
-        Ok(())
+        Ok(PostDispatchInfo {
+            actual_weight: {
+                let acc_weight = WeightConsumed::<T>::get();
+                Some((total_weight - acc_weight.weight_limit) + acc_weight.weight_limit)
```
[the remark this answers is not in the document]
Comment (Wizdave97) [reply pallet-ismp/src/lib.rs:554]: Typo, meant to add the used gas instead

File: Cargo.toml [hunk 7]
```diff
@@ -18,3 +19,6 @@ members = [
 
 [workspace.dependencies]
 enum-as-inner = "=0.5.1"
+
+[patch.crates-io]
+ruint = { git = "https://github.com/recmo/uint.git", rev = "e34200e114ac8fe93cf8e815e92601860ae4d7b8" }
```
Comment (Wizdave97) [inline Cargo.toml:24]: Patch until the next release of alloy-rs

**Review by seunlanlege — APPROVED** [review]
boom

**Pull request merged**